pkg/typedef/typedef.go

```diff
@@ -197,32 +216,49 @@ const (
 	CacheArrayLen
 )
 
-func prettyCQL(query string, values Values, types Types) string {
-	if len(values) == 0 {
-		return query
+func prettyCQL(builder *bytes.Buffer, q string, values Values, types []Type) error {
+	builder.Grow(len(q))
+
+	if len(types) == 0 {
+		builder.WriteString(q)
+		return nil
 	}
 
-	k := 0
-	out := make([]string, 0, len(values)*2)
-	queryChunks := strings.Split(query, "?")
-	out = append(out, queryChunks[0])
-	qID := 1
-	for _, typ := range types {
-		tupleType, ok := typ.(*TupleType)
-		if !ok {
-			out = append(out, typ.CQLPretty(values[k]))
-			out = append(out, queryChunks[qID])
-			qID++
-			k++
+	var (
+		skip int
+		idx  int
+	)
+
+	for pos, i := strings.Index(q[idx:], "?"), 0; pos != -1; pos = strings.Index(q[idx:], "?") {
+		str := q[idx : idx+pos]
+		// Just to skip the ? in the query, happens only in TUPLE TYPES
+		if skip > 0 {
+			skip--
+			idx += pos + 1
 			continue
 		}
-		for _, t := range tupleType.ValueTypes {
-			out = append(out, t.CQLPretty(values[k]))
-			out = append(out, queryChunks[qID])
-			qID++
-			k++
+
+		builder.WriteString(str)
+
+		var value any
+
+		switch tt := types[i].(type) {
+		case *TupleType:
+			skip = tt.LenValue()
+			value = values[i : i+skip]
+			values = values[skip:]
+		default:
+			value = values[i]
```

Comment:
If you get less values than expected it is going to panic

Reply:
To me in both cases is ok to panic, I can special case it further, but if it panics here, that means something is wrong with generation (which it is). But after many many runs, it never fails here, it fails on types not being good but never on offset (at least for now)

Reply:
I did the check above on line 228.